{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fd5b96c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "62b65435",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pip install names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b492745e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import names\n",
    "import random\n",
    "from random import randint,randrange\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4b200a73",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a folder to store generated data\n",
    "directory=\"generated_json\"\n",
    "if not os.path.isdir(directory):\n",
    "    os.mkdir(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "001e96da",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define the number of JSON files that will be generated\n",
    "num_files = 10\n",
    "\n",
    "#Set up the template JSON file\n",
    "template = {\n",
    "   \"patient_info\":{\n",
    "      \"Patient\": \"\",\n",
    "      \"Specimen\": \"\",\n",
    "      \"Referral_Lab\": \"\",\n",
    "      \"URN\": \"\",\n",
    "      \"DOB\": \"\",\n",
    "      \"Sex\": \"\",\n",
    "      \"Lab_No\": \"\",\n",
    "      \"Ext_Ref\": \"\",\n",
    "      \"Collected\": \"\",\n",
    "      \"Received\": \"\",\n",
    "      \"Requester\": \"\"\n",
    "   },\n",
    "   \"reportable_variants\": [\n",
    "   ],\n",
    "   \"other_info\":{\n",
    "   \"Clinical_Indication\":\"Ploycythaemia\",\n",
    "   \"Correlative_Morphology\":\"Not provided\",\n",
    "   \"Specimen_Details\":\"N/A\"\n",
    "   },\n",
    "   \"report_title\":{\n",
    "        \"Test_Description\":\"Somatic variant analysis of 22 genes with clinical significance in myeloproliferative neoplasms. Referto Panel Summary for gene list.\",\n",
    "        \"Result_Summary\":\"ASSUMED SOMATIC VARIANTS DETECTED. See Reportable Variants table below for details. \",\n",
    "        \"Clinical_Interpretation\":\"Variants in DNMT3A and SH2B3 were detected in this bone marrow aspirate sample consistent witha diagnosis of a myeloproliferative neoplasm (MPN) or idiopathic erythrocytosis. No typical MPNdriver variants (JAK2/CALR/MPL) were detected in this sample. Please correlate with morphological, immunophenotypic and cytogenetic features. Refer to final page of report for further information regarding the clinical utility of genomic testing inthis context.\",\n",
    "        \"reportable_variants_description\":\"Please note, variant origin (somatic or germline) cannot be determined by this assay. Variant origin is assumedhere based on ancillary information (e.g. population databases, literature, variant read frequency) for the purposeof clinical interpretation however testing of a germline sample may be recommended in some circumstances.\"\n",
    "        \n",
    "    },\n",
    "    \"Test_Methodology\": \n",
    "                \"DNA is analysed by targeted gene sequencing of coding regions and flanking splice sites (within 2 bp) of the genes listed below. Libraries areprepared using a custom QIAGEN QIAseq single primer extension-based panel (Peter MacCallum Cancer Centre AllHaem v1) and sequencedon an Illumina NextSeq500 with 150 bp paired end reads. A customised CLC bioinformatics pipeline including QIAGEN CLC enterprise solutionsis used to generate aligned reads and call variants (single nucleotide variants and short insertions or deletions) against the hg19 human referencgenome. Variants are analysed using PathOS software (Peter Mac) and described according to HGVS nomenclature version 19.01(http://varnomen.hgvs.org/) with minor differences in accordance with Peter MacCallum Cancer Centre Molecular Pathology departmental policy.The following population variation and cancer or genetic disease databases are commonly used in addition to literature review to assist withvariant interpretation: the Genome Aggregation Database (gnomAD; gnomad.broadinstitute.org), the Catalogue of Somatic Mutations in Cancer(COSMIC; cancer.sanger.ac.uk), ClinVar (ncbi.nlm.nih.gov/clinvar) and the IARC TP53 Database (p53.iarc.fr). Variant origin (i.e. somatic orgermline) is assumed based on ancillary information (e.g. population databases, literature, variant read frequency) for the purpose of clinicalinterpretation. All assumed somatic variants are reported (and generally considered clinically significant). Variants of uncertain origin are alsoreported, as are likely benign germline polymorphisms if sufficiently rare and otherwise undescribed. Testing of a non-haematological specimenmay be recommended to evaluate variant origin. Recurrent population variants are not reported. Somatic variant categorisation (modified fromAMP/ASCO/CAP guidelines1) â€“ Variants are curated and categorised according to the clinical context of the patient and categorised asDIAGNOSTIC (the variant either defines a diagnostic category or is sufficiently specific for the clinical context to contribute to diagnosticsubcategorisation), PROGNOSTIC (the variant has been associated in large trials/series with inferior or superior outcomes in either the contextof a specific therapy or independent of therapy. Note this does not take into account interaction between prognostic variants present in theindividual patient. Relevant pairwise interactions are presented in the clinical summary), DRUG TARGET (the variant or variant class is specificallytargeted by a therapeutic agent, this category only includes therapeutic agents that are clinically advanced and generally available through eitherreimbursement or clinical trials [i.e. not early stage investigational agents]), DRUG RESISTANCE (the variant is specifically associated withresistance to a targeted agent [i.e. does not include non-specific resistance to non-targeted therapies]), MRD MARKER (the variant is anestablished biomarker for which assessment at MRD sensitivity after therapy is accepted practice). If the variant is not categorised into any of the above categories it is assigned CLONAL MARKER indicating its utility in defining the presence of a clonal haematopoietic process in thespecimen. These categorisations are general in nature and may not be applicable to the specific clinicopathological context of the patient.\",\n",
    "    \"Test_Limitations\":\"The detection limit of this assay for specimens sequenced to the target read depth of 500x is a variant allele frequency (VAF) of approximately2% with the exception of ASXL1 c.1934dup;p.Gly646Trpfs*12 (detection limit ~ 5%). This assay is primarily qualitative however, the variant readfrequency (VRF) is provided to assist with variant interpretation and is assumed to approximate VAF in most instances (noting that the VAF ofsome insertions/deletions may be underrepresented due to assay-based allele bias). The measurement of uncertainty provided as a percentagerelative standard uncertainty (i.e. CV%) for variants with VAFs of 5%, 10%-20%, 30%-40% and 50% are on average, 10.2%, 10.4%, 3.5% and4.4%, respectively. Copy number variations, loss of heterozygosity, structural rearrangements or aneuploidies are not reported. Insertions ordeletions (particularly those > 25 bp in length), including FLT3-ITDs, are not reliably detected by this assay. Genes are analysed using thereference transcripts listed below; coding exons found in alternative transcripts are not assessed by this assay. This assay does not distinguishbetween somatic and germline variants. In addition, the clonal origin of somatic variants (i.e. disease compartment or cell lineage) cannot bedetermined. Synonymous variants are not routinely reported. Please note Peter Mac assumes sample identification, family relationships, andclinical diagnoses are as stated on the request. Our clinical recommendations may be based on evidence from third-party data sources and shouldbe interpreted in the context of all other clinical and laboratory information for this patient.\",\n",
    "    \"Panel_Summary\":{\n",
    "        \"Summary\":\"Please note variants may not be optimally detected in genes with less than 100% coverage. The gene coverage above is considered acceptablegiven the available information about the clinical context, however please contact the laboratory for further advice should specific genes coveredat less than 100% require full coverage. A list of regions with suboptimal coverage is available upon request.\",\n",
    "        \"Reported_by\":\"Dr Ing Soo Tiong (Consultant Haematologist)\",\n",
    "        \"Authorized_by\":\"A/Prof Piers Blombery (Consultant Haematologist)\",\n",
    "        \"Reported\":\"16-Mar-2023\"\n",
    "        \n",
    "    },\n",
    "    \"References\":\"1. Li MM, Datto M, Duncavage EJ, et al. Standards and Guidelines for the Interpretation and Reporting of Sequence Variants in Cancer:A Joint Consensus Recommendation of the Association for Molecular Pathology, American Society of Clinical Oncology, and College ofAmerican Pathologists. J Mol Diagn 2017; 19(1): 4-23.\"\n",
    "   \n",
    "}\n",
    "#Define the list for each of the element\n",
    "specimens = [\"Blood test\",\"Urine test\",\"Cerebrospinal fluid (CSF) test\",\"Pleural fluid test\",\"Muscle biopsy test\",\"Bone biopsy test\",\n",
    "             \"Pleural fluid test\",\"Synovial fluid test\",\"Ascites fluid test\",\"Amniotic fluid test\",\"skin biopsy test\",\n",
    "             \"Bronchoalveolar lavage (BAL) fluid test\",\"Peritoneal fluid test\",\"Tissue biopsy test\",\"Lymph node biopsy test\",\n",
    "             \"Bone biopsy test\",\"Fine needle aspiration (FNA) test\",\"Bone marrow aspirate\"]\n",
    "\n",
    "referral_labs = ['Alpha Pathology','Dorevitch Pathology','VCS Pathology','Dorevitch Pathology','4Cyte Pathology','Melbourne Pathology']\n",
    "\n",
    "URN = \"PMEX\"+str(randint(100000,999999))\n",
    "\n",
    "months = ['Jan','Fed','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec']\n",
    "\n",
    "dates = randint(1,10)\n",
    "\n",
    "years = randint(1960,2000)\n",
    "#To make sure the value has same format as the original one,here we add 0 if the date is smaller than 10\n",
    "if(dates<10):\n",
    "    dates = str(0)+str(dates)\n",
    "\n",
    "DOB = str(dates)+\"-\"+random.choice(months)+\"-\"+str(years)\n",
    "\n",
    "Collected = str(dates)+\"-\"+random.choice(months)+\"-\"+str(2022)\n",
    "#Here the received date must be later than the collected data\n",
    "Received = str(dates)+\"-\"+random.choice(months)+\"-\"+str(2023)\n",
    "\n",
    "Requester = \"Dr\"+names.get_full_name()\n",
    "\n",
    "sex = ['M','F']\n",
    "\n",
    "cells = ['Somatic','Germ','Stem','Cancer','Prokaryotic','Eukaryotic']\n",
    "\n",
    "Gene = ['DNMT3A','ASXL1','FLT3','SH2B3','TP53','CFTR','FOXP2','ALDH2','BRCA1','BRCA2']\n",
    "\n",
    "variant = ['c.2842del, p.(Glu948Argfs*3)','FLT3-ITD (allelic ratio 0.76)','c.2645G>A, p.(Arg882His)','c.1345G>A, p.(Glu449Lys)']\n",
    "\n",
    "AML = ['PROGNISTIC','PROGNOSTIC / DRUG TARGET','CLONAL MARKER','DIAGNOSTIC']\n",
    "\n",
    "\n",
    "\n",
    "#Using for loop to automatically generate JSON files\n",
    "for i in range(num_files):\n",
    "    \n",
    "    variantTimes = randint(0,5)\n",
    "    \n",
    "    data = template.copy()\n",
    "    \n",
    "    data['reportable_variants'] = []\n",
    "    \n",
    "    data['patient_info']['Patient'] = names.get_full_name()\n",
    "    \n",
    "    data['patient_info']['Specimen'] = random.choice(specimens)\n",
    "    \n",
    "    data['patient_info']['Referral_Lab'] = random.choice(referral_labs)\n",
    "    \n",
    "    data['patient_info']['URN'] = URN\n",
    "    \n",
    "    data['patient_info']['DOB'] = DOB\n",
    "    \n",
    "    data['patient_info']['Sex'] = random.choice(sex)\n",
    "    \n",
    "    data['patient_info']['Lab_No'] = str(randint(10000000,19999999))\n",
    "    \n",
    "    data['patient_info']['Ext_Ref'] = str(randint(100000000,9999999999))\n",
    "    \n",
    "    data['patient_info']['Collected'] = Collected\n",
    "    \n",
    "    data['patient_info']['Received'] = Received\n",
    "    \n",
    "    data['patient_info']['Requester'] = Requester\n",
    "    \n",
    "    \n",
    "    for j in range (0,variantTimes):\n",
    "        data['reportable_variants'].append(\n",
    "            {\n",
    "                 \"ASSUMED_ORIGIN\": random.choice(cells),\n",
    "                 \"GENE\": random.choice(Gene),\n",
    "                 \"VARIANT\": random.choice(variant),\n",
    "                 \"VRF\": str(randint(1,20)),\n",
    "                 \"CLINICAL_SIGNIFICANCE_IN_AML\": random.choice(AML)  \n",
    "            }\n",
    "        )\n",
    "        \n",
    "    \n",
    "    \n",
    "    #write json file,set the name as data(i).json\n",
    "    file = os.path.join(directory, f\"data{i}.json\")\n",
    "    with open(file,\"w\")as f:\n",
    "        json.dump(data,f,indent=3)\n",
    "    \n",
    "    \n",
    "        \n",
    "\n",
    "    \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
